design/one-pager-generic-datasource.md

```diff
@@ -0,0 +1,171 @@
+# Patch from Generic Datasources
+
+* Owner: Maximilian Blatt (maximilian.blatt@accenture.com, @mistermx)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## 1 Background
+
+Crossplane currently supports patching from composite resources to composed
+resources and vice versa. It is also possible to copy values between two
+composed resources inside one compositiion by patching the value from one
+composed back to the composite and then to the second composed resource.
+
+However, Crossplane currently does not provide a way to patch values from
+dynamic sources. Compositions are pratically static right now.
+
+## 2 Current workarounds
+
+While compositions itself are not dynamic, this restrictions can be circumvented
+using one of the following methods:
+
+* `provider-helm`: using [`spec.forProvider.valuesFrom.secretKeyRef`](https://doc.crds.dev/github.com/crossplane-contrib/provider-helm/helm.crossplane.io/Release/v1beta1@v0.10.0#spec-forProvider-valuesFrom-secretKeyRef)
+* `provider-kubernetes`: using [`spec.forProvider.references[].patchesFrom`](https://doc.crds.dev/github.com/crossplane-contrib/provider-kubernetes/kubernetes.crossplane.io/Object/v1alpha1@v0.3.0#spec-references-patchesFrom)
+
+However, none of these methods are Crossplane native and are not available to
+every provider. Furthermore, they would require an additional MR (here `Object`
+or `Release`) to deploy the composed resource increasing the complexity of the
+composition.
+
+## 3 Goals
+
+The goal of this document is to define a proposal for extracting environment
+specific data considering the following requirements:
+
+1. Allow the creation of generic compositions that are able to load environment 
+specific parameters during compose time.
+2. Allow users to define these values on cluster level.
+3. Prevent users from escalting their rights on the cluster and read out fields
+from objects they shouldn't have access to.
+4. (Bonus) Allow package writers to include default values in configuration
+packages.
+
+## 4 Proposed Solution
+
+Introduce a new Crossplane CRD `ValueSet` that can store arbitrary data similar
+to K8s `ConfigMaps` but supports more complex data types:
+
+```yaml
+---
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: ValueSet
```

Comment:
I think it's always good to keep the naming close to the upstream so that people would know what it does immediately. I'm wondering if we can come up with something closer to upstream rather than a complete new name.

The concept is very similar to `ConfigMap` except that it's common to the whole cluster and every composition author can access it. I don't have any suggestion that I feel strongly but some possible ones could be `ConfigSet`, `CommonConfigMap`, `ConfigMap` (same name, its own group), `CommonMap`, `ClusterConfigMap`.

Reply:
I thought `ValueSet` to be a good choice because it just contains an arbitrary set of values defined by the user. Other than that `ConfigSet` would work for me as well.